### ISSUE TITLE
Miscellaneous UI improvements

### ISSUE THREAD

**Issue by BWagener** [bohning/usdb_syncer#69]
I wanted to start a discussion on UI improvements.
There are two things I have noticed so far that could be improved in my opinion:

- main window size exceeds smaller screens
  - this is how the syncer looks on a 1280x800 screen (full screen): 
![image](https://user-images.githubusercontent.com/19321042/210140879-ffd63824-2080-4fca-b9af-71a44801f80b.png)
  - this is how the syncer looks on a 1280x800 screen (docked to the side - half screen as a workaround): 
![image](https://user-images.githubusercontent.com/19321042/210140975-a5003901-51e1-479e-b6e3-0d698d7af0a3.png)
  - this is how it looks with an [attempted improvement](https://github.com/bohning/usdb_syncer/compare/main...BWagener:usdb_syncer:fix_filter_width): ![image](https://user-images.githubusercontent.com/19321042/210140784-846308e9-efab-4271-b5c9-7b90bd1e063d.png)
    - I have assigned a fixed width to the filter combo boxes to achieve this, this is not a nice solution but mitigates the combo boxes always assuming the width of their longest item, which are quite long for the artist and title boxes. This wouldn't look nice on higher resolutions.
- the song table exceeds the main window (see third screenshot)
  - the table columns assume the width of their longest element if their header is set to `QHeaderView.ResizeMode.Interactive`, which due to a few very long artist and title names makes the table exceed the main window
  - a solution to this could be defaulting the interactive columns to assume a fraction of the table parent's width (minus the fixed columns width). At least up to a minimum width, below which the table should still be allowed to exceed it's parent horizontally
  - another nice addition on this front would be to remember user configured column width and default to those if they exist, I found [this approach](https://www.qtcentre.org/threads/26675-QTableView-saving-restoring-columns-widths) to this

Let me know what you think about these points and any you have to add! I understand if certain improvements or additions are too early at this point in development but it still might be worth it to collect them and maybe split more important ones off into separate issues.

**Comment by RumovZ** [comment]
Good points! On WSL, the app even crashes, because we try to open it in fullscreen, which doesn't fit.

I think the table is not that much of a problem, because it is a scroll area. But I agree that we waste a lot of space by adjusting to the longest entry. We probably don't need to worry too much about defaults if we remember user resizings. I would only hard code the width of the checkmark columns, which we statically know the maximum width of.
There are more columns we can potentially add in the future, so the ability to toggle columns may also be valuable.

The line of filters is presuambly what can crash the app. Not sure what we should use here that looks alright on narrow and wide devices and is potentially extensible in the future.

There is more stuff currently off with the UI. A lot of things have been added or removed without taking into account the overall layout. Although I assume things will continue to change frequently for some time, I would love to see someone starting to improve things!

**Comment by BWagener** [comment]
Cool!

I'm going to try to find a solution for the filters. There's two approaches on my mind at the moment.
- programmatically set their width relative to the available horizontal space, [this](https://forum.qt.io/topic/14636/how-to-set-the-size-percentage-of-a-qwidget/2) seems like a good approach two have the filters share a fraction of their parent's width
- allow the filter "row" to overflow to more "rows", I'm more keen on the first solution but this one scales with however many columns there might be in the future

Concerning the filter(s) for the presence of txt, audio, video etc. - those are not implemented yet correct? Is there already a desired solution? At the moment there is one combo box for the files - should this contain all the possible combinations of filter options for the 5 file types (this seems like a lot)? Or one combo box per file type? Or maybe open a submenu?
I don't have a strong opinion here but can see a submenu with one item per file type look nice:
```
Files
    > Text
    |   > Any
    |    Yes
    |      No
    Audio
    Video
    ...
```

In regards to the table, I agree on the fixed width for checkmark columns and the defaults for other columns not being so important if we can persist widths configured by the user 👍

To summarize I can see three items of work:
- find a solution for the filter widths
- implement filters for the existence of the 5 file types
- persist column widths of the table

What do you think? Should any of these be moved to a separate issue?

**Comment by RumovZ** [comment]
> programmatically set their width relative to the available horizontal space, [this](https://forum.qt.io/topic/14636/how-to-set-the-size-percentage-of-a-qwidget/2) seems like a good approach two have the filters share a fraction of their parent's width

I wonder how well this would work when resizing the window. 🤔 Also, as the widgets cannot be shrunk indefinitely, the filter area would still need to be scrollable, wouldn't it?

While overflowing / wrapping rows sound great, another solution could be to arrange the filters vertically. They could be placed next to the table. While that would take up some horizontal space, I can imagine it would also be more intuitive to use.

> Concerning the filter(s) for the presence of txt, audio, video etc. - those are not implemented yet correct?

Correct. I started implementing that, but then had to work on the actual file synchronisation. I was planning to only add a filter for the .txt or .usdb file, because I'm not sure if adding all those other filfters are worth bloating the GUI.
I like your idea with the nested dropdown menu, but we'd need a way to display the active filters then, unless you can only ever have one file filter active at the same time.

Provided we had a way to display active filters separatly from the dropdown menus, we could go even further and move other filters into the same dropdown menu as well. So you would have a single button `Add Filter` and after selecting it you could choose between language, views, golden notes, files, etc. After you choose 'Golden Notes > Yes', a label with that text would appear below the fitler menus.

And since I'm already rambling, here is yet another idea. 😛 We could make the whole thing text based. Then we'd just need a text input and users could enter arbitrarily complex search expressions, e.g. `artist:Queen OR "artist:Freddie Mercury" AND golden:yes`. Then we'd need our own search syntax and parser, but we could expand it arbitrarily without being bound by the GUI. There could still be widgets to help with common simple search cases.
Only question is whether we need such a powerful search engine. 😅

**Comment by mjhalwa** [comment]
Just an idea, bit how about splitting information apart? We could split die UI in multiple (tabbed?) sections: `Song List | My Songs | Downloads`

Then
- `song list` ist just the list from usdb and might have a single additional column for downloaded "yes/no".
- `My Songs` and `Downloads` show just ID, Artist, Title, Edition and Language
- `Downloads` lists files until download is done. txt, audio and video column indicate download progress (see #17)
- `My Songs` lists songs when done downloading, txt, audio and video indicate locally available files

**Comment by BWagener** [comment]
> I wonder how well this would work when resizing the window.

I played around with this before writing the issue, but I could only test on a small screen. I've just tested this [initial attempt](https://github.com/BWagener/usdb_syncer/tree/filter-width-working-demo) on a larger screen and the filters actually already do resize to fill the available space (shared in equal parts). It works fine with resizing as well as far as I can see.

> 🤔 Also, as the widgets cannot be shrunk indefinitely, the filter area would still need to be scrollable, wouldn't it?

Yes, although if we were to opt for supporting screens up to a minimal width this solution could serve the purpose well enough.

> While overflowing / wrapping rows sound great, another solution could be to arrange the filters vertically. They could be placed next to the table. While that would take up some horizontal space, I can imagine it would also be more intuitive to use.

Personally I prefer the horizontal layout for the filters as opposed to a vertical one.

> Provided we had a way to display active filters separatly from the dropdown menus, we could go even further and move other filters into the same dropdown menu as well. So you would have a single button `Add Filter` and after selecting it you could choose between language, views, golden notes, files, etc. After you choose 'Golden Notes > Yes', a label with that text would appear below the fitler menus.

I really like this approach! I would like to start playing around with this attempt first.

> I was planning to only add a filter for the .txt or .usdb file, because I'm not sure if adding all those other filfters are worth bloating the GUI.

I find it useful to have the ability to for example subselect all local songs that lack an audio file - so far I have used the sorting to have a workable result but filters would be ideal for this use-case.

> And since I'm already rambling, here is yet another idea. 😛 We could make the whole thing text based. Then we'd just need a text input and users could enter arbitrarily complex search expressions, e.g. `artist:Queen OR "artist:Freddie Mercury" AND golden:yes`. Then we'd need our own search syntax and parser, but we could expand it arbitrarily without being bound by the GUI. There could still be widgets to help with common simple search cases. Only question is whether we need such a powerful search engine. 😅

The programmer's dream 😄. I think this goes in a direction that could be adapted to work with @mjhalwa's [idea](https://github.com/bohning/usdb_syncer/issues/61#issuecomment-1368113611).

I think there are use-cases for facilitating searches such as the example you gave but I believe a good first effort could be to go with individual filters that can be added using an  `Add Filter` button as you described. The generic search field could always be extended with the aforementioned search parser to build custom queries!

**Comment by RumovZ** [comment]
> Just an idea, bit how about splitting information apart? We could split die UI in multiple (tabbed?) sections: Song List | My Songs | Downloads

Also sounds interesting, but is this really different to having a combined song list with a 'downloaded' filter? We could add a 'downloading' filter as well, but I think we should track active downloads in a separate list, because they change fast and the table widget is not really made for rapidly changing data.

@BWagener, all makes sense. I'm excited to see what you'll come up with!

**Comment by BWagener** [comment]
> @BWagener, all makes sense. I'm excited to see what you'll come up with!

I played around with the idea yesterday but I didn't get past the struggles I had with qt Designer. It will take me some time to adapt to this tool :D

**Comment by BWagener** [comment]
@RumovZ @mjhalwa 
Seeing how I've been unable to get started on the UI work for the past three weeks I want to suggest at least sharing the minimal improvements regarding the window width I have completed during the holidays on this branch https://github.com/BWagener/usdb_syncer/tree/fix_filter_width.
I will rebase the work and open a PR to see if you agree with the minimal adjustments I made.

Cheers

**Comment by RumovZ** [comment]
Closing as it seems all points have been resolved by #94 or subsequent changes.
